kim_cnn/train.py

```diff
@@ -0,0 +1,181 @@
+print("LABELS:",LABEL.vocab.itos)
+print("Train instance", len(train))
+print("Dev instance", len(dev))
+print("Test instance", len(test))
+
+
+if args.resume_snapshot:
+    if args.cuda:
+        model = torch.load(args.resume_snapshot, map_location=lambda storage, location: storage.cuda(args.gpu))
+    else:
+        model = torch.load(args.resume_snapshot, map_location=lambda storage, location: storage)
+else:
+    model = KimCNN(config)
+    if args.cuda:
+        model.cuda(args.gpu)
+        print("Shift model to GPU")
+
+
+embedding = Embedding(config, TEXT.vocab.vectors)
+
+
+parameter = filter(lambda p: p.requires_grad, model.parameters())
+#for idx, p in enumerate(parameter):
+#    print(idx, p)
+optimizer = torch.optim.Adadelta(parameter, lr=args.lr, weight_decay=args.weight_decay)
+criterion = nn.CrossEntropyLoss()
+early_stop = False
+best_dev_acc = 0
+iterations = 0
+iters_not_improved = 0
+epoch = 0
+start = time.time()
+header = '  Time Epoch Iteration Progress    (%Epoch)   Loss   Dev/Loss     Accuracy  Dev/Accuracy'
+dev_log_template = ' '.join('{:>6.0f},{:>5.0f},{:>9.0f},{:>5.0f}/{:<5.0f} {:>7.0f}%,{:>8.6f},{:8.6f},{:12.4f},{:12.4f}'.split(','))
+log_template = ' '.join('{:>6.0f},{:>5.0f},{:>9.0f},{:>5.0f}/{:<5.0f} {:>7.0f}%,{:>8.6f},{},{:12.4f},{}'.split(','))
+os.makedirs(args.save_path, exist_ok=True)
+os.makedirs(os.path.join(args.save_path, args.dataset), exist_ok=True)
+print(header)
+
+
+while True:
+    if early_stop:
+        print("Early Stopping. Epoch: {}, Best Dev Acc: {}".format(epoch, best_dev_acc))
+        break
+    epoch += 1
+    train_iter.init_epoch()
+    n_correct, n_total = 0, 0
+
+    for batch_idx, batch in enumerate(train_iter):
+        # Batch size : (Sentence Length, Batch_size)
+        iterations += 1
+        model.train(); optimizer.zero_grad()
+        #print("Text Size:", batch.text.size())
+        #print("Label Size:", batch.label.size())
+
+        model_input = embedding.embed(batch)
+        scores = model(model_input)
+        n_correct += (torch.max(scores, 1)[1].view(batch.label.size()).data == batch.label.data).sum()
+        n_total += batch.batch_size
+        train_acc = 100. * n_correct / n_total
+
+        loss = criterion(scores, batch.label)
+        loss.backward()
+
+        optimizer.step()
+
+
+        # Evaluate performance on validation set
+        if iterations % args.dev_every == 1:
+            # switch model into evalutaion mode
+            model.eval(); dev_iter.init_epoch()
+            n_dev_correct = 0
+            dev_losses = []
+            for dev_batch_idx, dev_batch in enumerate(dev_iter):
+                model_input = embedding.embed(dev_batch)
+                scores = model(model_input)
+                n_dev_correct += (torch.max(scores, 1)[1].view(dev_batch.label.size()).data == dev_batch.label.data).sum()
+                dev_loss = criterion(scores, dev_batch.label)
+                dev_losses.append(dev_loss.data[0])
```

Comment:
adapt the torch code to PyTorch 0.4
- dev_loss.data[0] should be modified to dev_loss.item()
refer to https://github.com/castorini/Castor/blob/master/common/trainers/sst_trainer.py#L46 for more info